File: funnyvote/app/src/main/java/com/heaton/funnyvote/ui/search/SearchItemAdapter.java
package com.heaton.funnyvote.ui.search;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.ImageView;
import android.widget.TextView;

import com.heaton.funnyvote.R;
import com.heaton.funnyvote.database.VoteData;

import java.util.List;


/**
 * Created by heaton on 2017/1/22.
 */

public class SearchItemAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    public static final int ITEM_TYPE_VOTE = 41;
    public static final int ITEM_TYPE_RELOAD = 42;
    public static final int ITEM_TYPE_NO_VOTE = 43;

    private boolean showReload;

    private List<VoteData> voteList;
    private Context context;
    private long maxCount;
    private SearchFragment.VoteSearchItemListener itemListener;

    public SearchItemAdapter(Context context
            , List<VoteData> datas, SearchFragment.VoteSearchItemListener itemListener) {
        this.context = context;
        this.itemListener = itemListener;
        this.voteList = datas;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == ITEM_TYPE_VOTE) {
            View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_list_search, parent, false);
            return new VHSearchItem(v, itemListener);
        } else if (viewType == ITEM_TYPE_RELOAD) {
            View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.card_view_item_reload, parent, false);
            return new ReloadViewHolder(v, itemListener);
        } else if (viewType == ITEM_TYPE_NO_VOTE) {
            View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.card_view_item_no_vote, parent, false);
            return new VHNoVote(v);
        }
        return null;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof VHSearchItem) {
            ((VHSearchItem) holder).setLayout(voteList.get(position));
        }
    }

    public void setVoteList(List<VoteData> voteList) {
        this.voteList = voteList;
    }

    @Override
    public int getItemCount() {
        if (voteList.size() > 0 && voteList.size() < maxCount) {
            showReload = true;
        } else {
            showReload = false;
        }
        if (showReload) {
            return voteList.size() + 1;
        } else {
            if (voteList.size() == 0) {
                return 1;
            } else {
                return voteList.size();
            }
        }
    }

    @Override
    public int getItemViewType(int position) {
        if (showReload && position == getItemCount() - 1) {
            return ITEM_TYPE_RELOAD;
        } else {
            if (voteList.size() == 0) {
                return ITEM_TYPE_NO_VOTE;
            } else {
                return ITEM_TYPE_VOTE;
            }
        }
    }

    public void setMaxCount(long count) {
        maxCount = count;
    }


    private class ReloadViewHolder extends RecyclerView.ViewHolder {
        public ImageView reloadImage;


        public ReloadViewHolder(View itemView, final SearchFragment.VoteSearchItemListener itemListener) {
            super(itemView);
            reloadImage = (ImageView) itemView.findViewById(R.id.img_load_more);
            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Animation animation = AnimationUtils.loadAnimation(context, R.anim.reload_rotate);
                    v.findViewById(R.id.img_load_more).startAnimation(animation);
                    itemListener.onReloadVote();
                }
            });
        }
    }


    private class VHNoVote extends RecyclerView.ViewHolder {

        public ImageView imgAddVote;
        public ImageView imgRefreshVote;
        public ImageView imgLogo;
        public TextView txtNoVote;

        public VHNoVote(View itemView) {
            super(itemView);
            imgAddVote = (ImageView) itemView.findViewById(R.id.imgAddVote);
            imgRefreshVote = (ImageView) itemView.findViewById(R.id.imgRefreshVote);
            imgLogo = (ImageView) itemView.findViewById(R.id.imgLogo);
            txtNoVote = (TextView) itemView.findViewById(R.id.txtNoVote);
            imgAddVote.setVisibility(View.GONE);
            imgRefreshVote.setVisibility(View.GONE);
            imgLogo.setVisibility(View.VISIBLE);
            txtNoVote.setText(R.string.wall_item_no_vote_search);
        }
    }
}

File: funnyvote/app/src/main/java/com/heaton/funnyvote/ui/search/SearchItemAdapter.kt
package com.heaton.funnyvote.ui.search

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import android.widget.ImageView
import android.widget.TextView
import com.heaton.funnyvote.R
import com.heaton.funnyvote.database.VoteData


/**
 * Created by heaton on 2017/1/22.
 */

class SearchItemAdapter(private val context: Context
                        , private var voteList: List<VoteData>
                        , private val itemListener: SearchFragment.VoteSearchItemListener
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private var showReload: Boolean = false
    private var maxCount: Long = 0

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        when (viewType) {
            ITEM_TYPE_VOTE -> {
                val v = LayoutInflater.from(parent.context).inflate(R.layout.item_list_search, parent, false)
                return VHSearchItem(v, itemListener)
            }
            ITEM_TYPE_RELOAD -> {
                val v = LayoutInflater.from(parent.context).inflate(R.layout.card_view_item_reload, parent, false)
                return ReloadViewHolder(v, itemListener)
            }
            ITEM_TYPE_NO_VOTE -> {
                val v = LayoutInflater.from(parent.context).inflate(R.layout.card_view_item_no_vote, parent, false)
                return VHNoVote(v)
            }
            else -> {
                return VHNoVote(parent)
            }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is VHSearchItem) {
            holder.setLayout(voteList[position])
        }
    }

    fun setVoteList(voteList: List<VoteData>) {
        this.voteList = voteList
    }

    override fun getItemCount(): Int {
        showReload = voteList.size in 1..(maxCount - 1)
        return if (showReload) {
            voteList.size + 1
        } else {
            if (voteList.isEmpty()) {
                1
            } else {
                voteList.size
            }
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (showReload && position == itemCount - 1) {
            ITEM_TYPE_RELOAD
        } else {
            if (voteList.isEmpty()) {
                ITEM_TYPE_NO_VOTE
            } else {
                ITEM_TYPE_VOTE
            }
        }
    }

    fun setMaxCount(count: Long) {
        maxCount = count
    }


    private inner class ReloadViewHolder(itemView: View, itemListener: SearchFragment.VoteSearchItemListener) : RecyclerView.ViewHolder(itemView) {
        var reloadImage: ImageView = itemView.findViewById<View>(R.id.img_load_more) as ImageView


        init {
            itemView.setOnClickListener { v ->
                val animation = AnimationUtils.loadAnimation(context, R.anim.reload_rotate)
                v.findViewById<View>(R.id.img_load_more).startAnimation(animation)
                itemListener.onReloadVote()
            }
        }
    }


    private inner class VHNoVote(itemView: View) : RecyclerView.ViewHolder(itemView) {

        var imgAddVote: ImageView = itemView.findViewById<View>(R.id.imgAddVote) as ImageView
        var imgRefreshVote: ImageView = itemView.findViewById<View>(R.id.imgRefreshVote) as ImageView
        var imgLogo: ImageView = itemView.findViewById<View>(R.id.imgLogo) as ImageView
        var txtNoVote: TextView = itemView.findViewById<View>(R.id.txtNoVote) as TextView

        init {
            imgAddVote.visibility = View.GONE
            imgRefreshVote.visibility = View.GONE
            imgLogo.visibility = View.VISIBLE
            txtNoVote.setText(R.string.wall_item_no_vote_search)
        }
    }

    companion object {
        const val ITEM_TYPE_VOTE = 41
        const val ITEM_TYPE_RELOAD = 42
        const val ITEM_TYPE_NO_VOTE = 43
    }
}
